File: src/asas_mcp/tools/scripts/ghidra_export.py
# Ghidra Headless Script: Decompile and Export Functions with Addresses
# This script runs inside Ghidra's Jython environment

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import json
import os

def run():
    program = currentProgram
    name = program.getName()
    print("Analyzing program: " + name)
    
    iface = DecompInterface()
    iface.openProgram(program)
    
    results = []
    
    fm = program.getFunctionManager()
    funcs = fm.getFunctions(True)
    
    for func in funcs:
        func_name = func.getName()
        func_addr = str(func.getEntryPoint())
        
        # Decompile
        res = iface.decompileFunction(func, 30, ConsoleTaskMonitor())  # 30s per function
        code = ""
        if res.decompileCompleted():
            decomp = res.getDecompiledFunction()
            if decomp:
                code = decomp.getC()
        
        results.append({
            "name": func_name,
            "address": func_addr,
            "code": code
        })
            
    # Save output
    try:
        output_path = os.environ.get("GHIDRA_OUTPUT_PATH", "/tmp/ghidra_output.json")
        with open(output_path, 'w') as f:
            json.dump(results, f)
        print("Success: Exported " + str(len(results)) + " functions to " + output_path)
    except Exception as e:
        print("Error saving results: " + str(e))

if __name__ == "__main__":
    run()

File: src/asas_mcp/tools/scripts/GhidraExport.java
// Ghidra Headless Script: Decompile and Export Functions with Addresses
// This script runs inside Ghidra's Java environment
// @category CTF-ASAS

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompiledFunction;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.FileWriter;
import java.util.Iterator;

public class GhidraExport extends GhidraScript {
    
    @Override
    public void run() throws Exception {
        println("CTF-ASAS Ghidra Export starting...");
        
        DecompInterface iface = new DecompInterface();
        iface.openProgram(currentProgram);
        
        FunctionManager fm = currentProgram.getFunctionManager();
        Iterator<Function> funcs = fm.getFunctions(true).iterator();
        
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        boolean first = true;
        int count = 0;
        
        while (funcs.hasNext()) {
            Function func = funcs.next();
            String name = func.getName();
            String addr = func.getEntryPoint().toString();
            
            // Decompile with 30s timeout per function
            DecompileResults res = iface.decompileFunction(func, 30, new ConsoleTaskMonitor());
            String code = "";
            if (res.decompileCompleted()) {
                DecompiledFunction df = res.getDecompiledFunction();
                if (df != null) {
                    code = df.getC();
                }
            }
            
            if (!first) sb.append(",");
            first = false;
            
            sb.append("{");
            sb.append("\"name\":\"").append(escapeJson(name)).append("\",");
            sb.append("\"address\":\"").append(escapeJson(addr)).append("\",");
            sb.append("\"code\":\"").append(escapeJson(code)).append("\"");
            sb.append("}");
            count++;
        }
        
        sb.append("]");
        
        String outputPath = System.getenv("GHIDRA_OUTPUT_PATH");
        if (outputPath == null || outputPath.isEmpty()) {
            outputPath = "/tmp/ghidra_output.json";
        }
        
        FileWriter fw = new FileWriter(outputPath);
        fw.write(sb.toString());
        fw.close();
        
        println("Success: Exported " + count + " functions to " + outputPath);
    }
    
    private String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
}
